File: javascript/sdk/otp.js
/**
 * OTP module - OTP request and verification
 * Equivalent to PHP send_otp.php and verify_otp.php
 */

const axios = require('axios');

class OTPService {
  /**
   * @param {string} applicationId - Application ID
   * @param {string} password - Application password
   * @param {string} baseURL - BDApps base URL
   */
  constructor(applicationId, password, baseURL = 'https://developer.bdapps.com') {
    this.applicationId = applicationId;
    this.password = password;
    this.baseURL = baseURL;
  }

  /**
   * Request OTP for a mobile number
   * @param {string} userMobile - Mobile number (without country code prefix, e.g., '01847026XX')
   * @param {object} [metaData] - Optional application meta data
   * @returns {Promise<object>} Response with referenceNo
   */
  async sendOtp(userMobile, metaData = null) {
    // Format mobile number with tel:88 prefix
    const subscriberId = userMobile.startsWith('tel:')
      ? userMobile
      : `tel:88${userMobile}`;

    const requestData = {
      applicationId: this.applicationId,
      password: this.password,
      subscriberId,
      applicationHash: metaData?.applicationHash || 'App Name',
      applicationMetaData: {
        client: metaData?.client || 'MOBILEAPP',
        device: metaData?.device || 'Samsung S10',
        os: metaData?.os || 'android 8',
        appCode: metaData?.appCode || 'https://play.google.com/store/apps/details?id=lk.dialog.megarunlor',
      },
    };

    const url = `${this.baseURL}/subscription/otp/request`;

    try {
      const response = await axios.post(url, requestData, {
        headers: { 'Content-Type': 'application/json' },
        httpsAgent: new (require('https').Agent)({ rejectUnauthorized: false }),
      });

      return response.data;
    } catch (error) {
      if (error.response) return error.response.data;
      throw error;
    }
  }

  /**
   * Verify OTP
   * @param {string} referenceNo - Reference number from sendOtp response
   * @param {string} otp - One-time password to verify
   * @returns {Promise<object>} Response with subscriptionStatus
   */
  async verifyOtp(referenceNo, otp) {
    const requestData = {
      applicationId: this.applicationId,
      password: this.password,
      referenceNo,
      otp,
    };

    const url = `${this.baseURL}/subscription/otp/verify`;

    try {
      const response = await axios.post(url, requestData, {
        headers: { 'Content-Type': 'application/json' },
        httpsAgent: new (require('https').Agent)({ rejectUnauthorized: false }),
      });

      return response.data;
    } catch (error) {
      if (error.response) return error.response.data;
      throw error;
    }
  }
}

module.exports = OTPService;

File: python/sdk/otp.py
"""
OTP module - OTP request and verification.
Equivalent to PHP send_otp.php and verify_otp.php.
"""

from typing import Optional

import httpx


class OTPService:
    """OTP request and verification service."""

    def __init__(
        self,
        application_id: str,
        password: str,
        base_url: str = "https://developer.bdapps.com",
    ):
        self.application_id = application_id
        self.password = password
        self.base_url = base_url

    async def send_otp(
        self, user_mobile: str, meta_data: Optional[dict] = None
    ) -> dict:
        """
        Request OTP for a mobile number.

        Args:
            user_mobile: Mobile number (without country code prefix, e.g., '01847026XX').
            meta_data: Optional application metadata dict.

        Returns:
            Response dict with referenceNo on success.
        """
        # Format mobile number with tel:88 prefix
        subscriber_id = (
            user_mobile
            if user_mobile.startswith("tel:")
            else f"tel:88{user_mobile}"
        )

        meta = meta_data or {}
        request_data = {
            "applicationId": self.application_id,
            "password": self.password,
            "subscriberId": subscriber_id,
            "applicationHash": meta.get("applicationHash", "App Name"),
            "applicationMetaData": {
                "client": meta.get("client", "MOBILEAPP"),
                "device": meta.get("device", "Samsung S10"),
                "os": meta.get("os", "android 8"),
                "appCode": meta.get(
                    "appCode",
                    "https://play.google.com/store/apps/details?id=lk.dialog.megarunlor",
                ),
            },
        }

        url = f"{self.base_url}/subscription/otp/request"

        try:
            async with httpx.AsyncClient(verify=False, timeout=30.0) as client:
                response = await client.post(
                    url,
                    json=request_data,
                    headers={"Content-Type": "application/json"},
                )
                return response.json()
        except Exception as e:
            return {"statusCode": "E1601", "statusDetail": str(e)}

    async def verify_otp(self, reference_no: str, otp: str) -> dict:
        """
        Verify OTP.

        Args:
            reference_no: Reference number from send_otp response.
            otp: One-time password to verify.

        Returns:
            Response dict with subscriptionStatus on success.
        """
        request_data = {
            "applicationId": self.application_id,
            "password": self.password,
            "referenceNo": reference_no,
            "otp": otp,
        }

        url = f"{self.base_url}/subscription/otp/verify"

        try:
            async with httpx.AsyncClient(verify=False, timeout=30.0) as client:
                response = await client.post(
                    url,
                    json=request_data,
                    headers={"Content-Type": "application/json"},
                )
                return response.json()
        except Exception as e:
            return {"statusCode": "E1601", "statusDetail": str(e)}
